### PR TITLE
Allow octal literals as arguments

### PR DESCRIPTION
**Pull request by mpeterv**
Use same function for parsing numerical literals in arithmetic
expressions and as arguments of primitives, e.g. 'port'. Add some
error reporting to this function; previously lexer had to fallback
to address parsing on incorrect number literals.

This improves tcpdump compatibility.

### CODE REVIEW

**Comment by kbara** [comment]
Would you explain/give an example of how it improves tcpdump compatibility?

**Comment by mpeterv** [comment]
@kbara it's just that, octal arguments. E.g. `port 010` is supported in tcpdump. Of course that's very minor, the main benefit is improvement of error messages.

**Comment by kbara** [comment]
Thanks for the clarification. LGTM, and the extra parse tests look like a sufficient regression test. WDYT, @andywingo ?

File: src/pf/parse.lua [hunk 1]
```diff
@@ -24,12 +24,55 @@ local punctuation = set(
...
+
+local function lex_number(str, pos, base)
+   local res = 0
+   local i = pos
+   while i <= #str do
+      local chr = str:sub(i,i)
+      local n = tonumber(chr, base)
+      if n then
+         res = res * base + n
+         assert(res <= 0xffffffff, 'integer too large: '..res)
```
Comment (andywingo) [inline src/pf/parse.lua]: Is 0xffffffffffoo.com a valid hostname?  I would think so.  Probably want to move this assertion to the end, when we know we've seen a terminator and have a valid number.  WDYT?

**Comment by andywingo** [comment]
LGTM with the nit, please let me know if the nit is correct and if so when you have a fix.  Cheers :)

File: src/pf/parse.lua (hunk 1 above)
Comment (kbara) [inline src/pf/parse.lua]: Yep, both 0xfoo.com and 0xf00.com resolve; I ran them through 'host'.

Comment (andywingo) [inline src/pf/parse.lua]: And I think these pass the patch as given.  What doesn't pass is a sequence of numbers at the start of a hostname that when parsed as a number is greater than 2^32.

Comment (kbara) [inline src/pf/parse.lua]: Yep. Adding more 'f's in the middle retains validity as a domain name, so I'm confirming your observation.

**Comment by mpeterv** [comment]
@andywingo assertion has to be moved out completely because a number with a valid terminator still can be start of a hostname, e.g. `0xffffffffff-oo.com`.

I've pushed a simple fix which duplicates the assertion. We could also add optional `max_pos` parameter to lex_number and check number size only when it's not used or is equal to position of last digit. If you don't mind the current fix I could do that later along with some refactoring.

**Comment by andywingo** [comment]
LGTM.  Thank you @mpeterv :)

**Pull request merged**